### PR TITLE
Update deploying-the-sample-app.md

### PR DESCRIPTION
**Pull request by dsmohan**
**Purpose**
Issues:
1. There seems to be some bug in the existing webapps (**saml2-web-app-pickup-dispatch** and saml2-web-app-pickup-manager) shared in the documentation. As after deploying and running the webapps, when trying to login, the authentication request fails due to signature verification failure.
2. After the deployment of webapps, in the sso.properties file under <TOMCAT_HOME>/apache-tomcat-<version>/webapps/webapp/WEB-INF/classes, the **SAML2.AssertionConsumerURL** has hostname as localhost.com whereas it should be localhost. Otherwise the flow won't work properly.
3. While configuring the service providers for these webapps in WSO2 IS Management Console, the **Assertion Consumer URL**  hostname is given as localhost.com in the document whereas it should be localhost.

The edits made in this document now resolves all the above 3 issues.

**Goals**
To provide solution to the above mentioned issues that will help in implementing the SSO with SAML use case successfully.

**Approach**
As the flow didn't work as expected with the given details in the document, I found hostname i.e. "localhost.com" could be changed to "localhost" in configurations. And for the authentication failure issue due to signature validation (an issue faced with another use-case earlier too), I gave a try by replacing the existing wso2carbon.jks file in the webapps with the WSO2 IS wso2carbon.jks as suggested by WSO2 support team. With these changes, the flow worked as expected.

**Test environment**
OS - Windows 64-bit
JDK- openjdk version "11.0.5"
WSO2 IS - 5.11.0
Apache Tomcat - 8.5.87
 
**Changes done in the document**
Added extra information in lines# 327, 328, 407, 408 and made minor edit in lines# 356, 437 for successful deployment and running of the saml2-web-app-pickup-dispatch and saml2-web-app-pickup-manager webapps to implement SSO with SAML successfully.
1. The lines# 327 and 407 provide resolution to the issue# 1 mentioned above.
2. Lines# 328 and 408 provide resolution to the issue# 2 mentioned above.
3. Changes in the lines# 356 and 437 provide resolution to the issue# 3 mentioned above.

### CODE REVIEW

File: en/docs/learn/deploying-the-sample-app.md [hunk 1]
```diff
@@ -324,6 +324,8 @@ Deploy this sample web app on a web container.
     `           <TOMCAT_HOME>/apache-tomcat-<version>/webapps          `
     .
 2.  Start the Tomcat server.
+3.  Copy wso2carbon.jks file from <IS_HOME>/repository/resources/security and replace it with the existing wso2carbon.jks file under the web application folder    <TOMCAT_HOME>/apache-tomcat-<version>/webapps/saml2-web-app-pickup-dispatch.com/WEB-INF/classes (If you miss this step, you will run into an authentication issue).
```
Comment (nilminiwso2) [inline en/docs/learn/deploying-the-sample-app.md:327]:
```suggestion
3.  Copy the `wso2carbon.jks` file from the `<IS_HOME>/repository/resources/security` folder and replace the existing `wso2carbon.jks` file in the  `<TOMCAT_HOME>/apache-tomcat-<version>/webapps/saml2-web-app-pickup-dispatch.com/WEB-INF/classes` folder.
     
     !!! info
          If you miss this step, you will run into an authentication issue.
```

File: en/docs/learn/deploying-the-sample-app.md [hunk 2]
```diff
@@ -324,6 +324,8 @@ Deploy this sample web app on a web container.
...
+4.  Edit sso.properties file present under <TOMCAT_HOME>/apache-tomcat-<version>/webapps/saml2-web-app-pickup-dispatch.com/WEB-INF/classes and change hostname from localhost.com to localhost in SAML2.AssertionConsumerURL
```
Comment (nilminiwso2) [inline en/docs/learn/deploying-the-sample-app.md:328]:
```suggestion
4.  Open the `sso.properties` file in the `<TOMCAT_HOME>/apache-tomcat-<version>/webapps/saml2-web-app-pickup-dispatch.com/WEB-INF/classes` folder and change the hostname in the SAML2 assertion consumer URL (`SAML2.AssertionConsumerURL` parameter) from `localhost.com` to `localhost` as shown below.

      ```bash
      SAML2.AssertionConsumerURL=http://localhost.com:8080/saml2-web-app-pickup-dispatch.com/home.jsp
      ```
```

File: en/docs/learn/deploying-the-sample-app.md [hunk 3]
```diff
@@ -402,6 +404,8 @@ Deploy this sample web app on a web container.
     `           <TOMCAT_HOME>/apache-tomcat-<version>/webapps          `
     .
 2.  Start the Tomcat server.
+3.  Copy wso2carbon.jks file from <IS_HOME>/repository/resources/security and replace it with the existing wso2carbon.jks file under the web application folder    <TOMCAT_HOME>/apache-tomcat-<version>/webapps/saml2-web-app-pickup-manager.com/WEB-INF/classes (If you miss this step, you will run into an authentication issue).
```
Comment (nilminiwso2) [inline en/docs/learn/deploying-the-sample-app.md:407]:
```suggestion
3.  Copy the `wso2carbon.jks` file from the `<IS_HOME>/repository/resources/security` folder and replace the existing `wso2carbon.jks` file in the  `<TOMCAT_HOME>/apache-tomcat-<version>/webapps/saml2-web-app-pickup-manager.com/WEB-INF/classes` folder.
     
     !!! info
          If you miss this step, you will run into an authentication issue.
```

File: en/docs/learn/deploying-the-sample-app.md [hunk 4]
```diff
@@ -402,6 +404,8 @@ Deploy this sample web app on a web container.
...
+4.  Edit sso.properties file present under <TOMCAT_HOME>/apache-tomcat-<version>/webapps/saml2-web-app-pickup-manager.com/WEB-INF/classes and change hostname from localhost.com to localhost in SAML2.AssertionConsumerURL
```
Comment (nilminiwso2) [inline en/docs/learn/deploying-the-sample-app.md:408]:
```suggestion
4.  Open the `sso.properties` file in the `<TOMCAT_HOME>/apache-tomcat-<version>/webapps/saml2-web-app-pickup-manager.com/WEB-INF/classes` folder and change the hostname in the SAML2 assertion consumer URL (`SAML2.AssertionConsumerURL` parameter) from `localhost.com` to `localhost` as shown below.

      ```bash
      SAML2.AssertionConsumerURL=http://localhost:8080/saml2-web-app-pickup-manager.com/home.jsp
      ```
```

**Review by dsmohan — COMMENTED** [review]
Reviewed

**Review by dsmohan — COMMENTED** [review]
Reviewed

**Review by dsmohan — COMMENTED** [review]
Reviewed

**Review by dsmohan — COMMENTED** [review]
Reviewed

File: en/docs/learn/deploying-the-sample-app.md [hunk 5]
```diff
@@ -324,6 +324,15 @@ Deploy this sample web app on a web container.
     `           <TOMCAT_HOME>/apache-tomcat-<version>/webapps          `
     .
 2.  Start the Tomcat server.
```
Comment (ThaminduDilshan) [inline en/docs/learn/deploying-the-sample-app.md:326]:
```suggestion
```

File: en/docs/learn/deploying-the-sample-app.md [hunk 6]
```diff
@@ -324,6 +324,15 @@ Deploy this sample web app on a web container.
...
+3.  Copy the `wso2carbon.jks` file from the `<IS_HOME>/repository/resources/security` folder and replace the existing `wso2carbon.jks` file in the  `<TOMCAT_HOME>/apache-tomcat-<version>/webapps/saml2-web-app-pickup-dispatch.com/WEB-INF/classes` folder.
+     
+     !!! info
+          If you miss this step, you will run into an authentication issue.
+4.  Open the `sso.properties` file in the `<TOMCAT_HOME>/apache-tomcat-<version>/webapps/saml2-web-app-pickup-dispatch.com/WEB-INF/classes` folder and change the hostname in the SAML2 assertion consumer URL (`SAML2.AssertionConsumerURL` parameter) from `localhost.com` to `localhost` as shown below.
+
+      ```bash
+      SAML2.AssertionConsumerURL=http://localhost:8080/saml2-web-app-pickup-dispatch.com/home.jsp
+      ```
```
Comment (ThaminduDilshan) [inline en/docs/learn/deploying-the-sample-app.md:335]:
```suggestion
2.  Copy the `wso2carbon.jks` file from the `<IS_HOME>/repository/resources/security` folder and replace the existing `wso2carbon.jks` file in the  `<TOMCAT_HOME>/apache-tomcat-<version>/webapps/saml2-web-app-pickup-dispatch.com/WEB-INF/classes` folder.
     
     !!! info
          If you miss this step, you will run into an authentication issue.
3.  Open the `sso.properties` file in the `<TOMCAT_HOME>/apache-tomcat-<version>/webapps/saml2-web-app-pickup-dispatch.com/WEB-INF/classes` folder and change the hostname in the SAML2 assertion consumer URL (`SAML2.AssertionConsumerURL` parameter) from `localhost.com` to `localhost` as shown below.

      ```bash
      SAML2.AssertionConsumerURL=http://localhost:8080/saml2-web-app-pickup-dispatch.com/home.jsp
      ```
4.  Start the Tomcat server.
```

File: en/docs/learn/deploying-the-sample-app.md [hunk 7]
```diff
@@ -402,6 +411,15 @@ Deploy this sample web app on a web container.
     `           <TOMCAT_HOME>/apache-tomcat-<version>/webapps          `
     .
 2.  Start the Tomcat server.
```
Comment (ThaminduDilshan) [inline en/docs/learn/deploying-the-sample-app.md:413]:
```suggestion
```

File: en/docs/learn/deploying-the-sample-app.md [hunk 8]
```diff
@@ -402,6 +411,15 @@ Deploy this sample web app on a web container.
...
+3.  Copy the `wso2carbon.jks` file from the `<IS_HOME>/repository/resources/security` folder and replace the existing `wso2carbon.jks` file in the  `<TOMCAT_HOME>/apache-tomcat-<version>/webapps/saml2-web-app-pickup-manager.com/WEB-INF/classes` folder.
+     
+     !!! info
+          If you miss this step, you will run into an authentication issue.
+4.  Open the `sso.properties` file in the `<TOMCAT_HOME>/apache-tomcat-<version>/webapps/saml2-web-app-pickup-manager.com/WEB-INF/classes` folder and change the hostname in the SAML2 assertion consumer URL (`SAML2.AssertionConsumerURL` parameter) from `localhost.com` to `localhost` as shown below.
+
+      ```bash
+      SAML2.AssertionConsumerURL=http://localhost:8080/saml2-web-app-pickup-manager.com/home.jsp
+      ```
```
Comment (ThaminduDilshan) [inline en/docs/learn/deploying-the-sample-app.md:422]:
```suggestion
2.  Copy the `wso2carbon.jks` file from the `<IS_HOME>/repository/resources/security` folder and replace the existing `wso2carbon.jks` file in the  `<TOMCAT_HOME>/apache-tomcat-<version>/webapps/saml2-web-app-pickup-manager.com/WEB-INF/classes` folder.
     
     !!! info
          If you miss this step, you will run into an authentication issue.
3.  Open the `sso.properties` file in the `<TOMCAT_HOME>/apache-tomcat-<version>/webapps/saml2-web-app-pickup-manager.com/WEB-INF/classes` folder and change the hostname in the SAML2 assertion consumer URL (`SAML2.AssertionConsumerURL` parameter) from `localhost.com` to `localhost` as shown below.

      ```bash
      SAML2.AssertionConsumerURL=http://localhost:8080/saml2-web-app-pickup-manager.com/home.jsp
      ```
4.  Start the Tomcat server.
```

**Comment by ThaminduDilshan** [comment]
> 1. There seems to be some bug in the existing webapps (saml2-web-app-pickup-dispatch and saml2-web-app-pickup-manager) shared in the documentation. As after deploying and running the webapps, when trying to login, the authentication request fails due to signature verification failure.

I couldn't reproduce any errors in the WSO2 IS 5.9.0 distribution with tomcat 8.x while trying the said samples with the `localhost.com`. Login flow worked without any errors.

However Replacing `wso2carbon.jks` can be considered as a valid step.

**Comment by malithie** [comment]
@ThaminduDilshan do we really need to have the URL as "localhost.com" ?
From an evaluation perspective it's an additional hop to configure the hostname.

Why not we go with "localhost" in docs which is the default URL and changing hostname is anyway covered separately.

**Comment by madurangasiriwardena** [comment]
Hi @dsmohan,

Thank you for your contributions.

As @ThaminduDilshan has mentioned above, the configs mentioned in the documentation is working without any issue. However, the requirement to use the localhost.com domain seems to be not necessary and an overhead to try out the sample.

Providing instructions to change the ACS URL in the config files is a much overhead compared to adding the host entry to the /etc/hosts. To simplify this, we need to change the default ACS URL provided in the sample application. But since it is not possible to update the sample application compatible with IS 5.9.0 now, we cannot accept this PR.

With the documentation refactoring done with IS 7.0.0, this `localhost.com` hostname seems to be no longer used in the upcoming IS 7.0.0 documentation. We will continue to evaluate this for the upcoming IS 7.0.0 release and fix if there are any issues.

**Comment by dsmohan** [comment]
Hi @madurangasiriwardena , @ThaminduDilshan ,

Thanks for the update.
What about the 1st issue - authentication failure due to signature validation which gets resolved by replacing the existing wso2carbon.jks file in the webapps with the one in the WSO2 IS?

Thanks,
Suman

**Pull request closed without merging**